### PR TITLE
Integration Tests Passing & Latency/LoggedOn Properties Added

### LINKED ISSUE #16: Add latency data
Latency will be an integer representing time the query took to complete.
It will also need to be added to the web app backend.

### LINKED ISSUE #18: Add latency property to logger middleware functionality
Project Queries in the web app will be updated to include this property once functionality is added to logger to calculate the timestamp (when the query was initially sent) and when the query request was responded back to.

It will be necessary to update the logger to include this property in the receiving of `res.locals.graphQLGate` as well as the sending of the data to the web app's backend.

### LINKED ISSUE #19: Add functionality to calculate the latency of queries
This latency should be calculated in the manipulation of `res.end`, taking the timestamp pulled from `res.locals.graphQLGate` and subtracting the current timestamp from it, calculating the amount in ms the query request took to complete and be responded back to.

### LINKED ISSUE #21: Coordinate the data being logged accross all  repositories
Data to be logged to the Gateway app database:
1. `timestamp`: number
2. `tokens`: number (tokens requested on the query)
3. `number`: number (quereyid)
4. `latency`: number
5. `complexity`: number
6. `depth`: number
7. `ip` address : string ??
8. `allowed` or rejected: boolean

- [ ] The `graphQL-gateway`schema, graphql types, types, should be updated with these fields. 
- [ ] `graphql-gate` should be adding the necessary information to res.locals for the logger to pickup
- [ ]  The `gate-logger` package should be logging the correct data
- [ ] The `graphql-gateway` front end should be fetching the correct data for display

### PR DESCRIPTION
**Pull request by jondeweydev**
### Summary

- Integration tests are passing and adding project queries to the web app's backend once test server sends mock response back to the 'client'.
- Queries now have the properties latency and loggedOn
- Functionality to get timestamp at beginning of middleware chain, loggedOn at end, and latency calculated between the two added.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

### Issues

closes #16 
closes #19 
closes #18 
closes #21 

### Evidence

Screen Shot 2022-06-30 at 6 08 54 PM
Screen Shot 2022-06-30 at 6 09 40 PM

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -29,32 +29,45 @@ const gateURI = 'http://localhost:3000';
...
-            // postQuery.post(req, res, next);
-            return res;
-        };
+    const newAuth = new AuthVerification(gateURI, projectID, apiKey);
+
+    const validate = newAuth.validation;
+    const verify = newAuth.verification;
+
+    // run the API Key verification process when gatelog is instantiated
+    validate().catch((err) => err);
```
Comment (shalarewicz) [inline src/index.ts]: These catches isn't really doing anything and the function will still return middleware regardless of whether the verification is successful. I think we should only return middleware if the API key verification is successful, if not we should throw an error letting the developer know the API key is invalid.

File: src/index.ts [hunk 2]
```diff
@@ -29,32 +29,45 @@ const gateURI = 'http://localhost:3000';
...
+        // runs logger's functionality upon response being sent back to client
+        res.on('finish', async (): Promise<void> => {
+            // calls timestamp of request's end
+            const loggedOn: number = new Date().valueOf();
+
+            // stores time between request's beginning and end
+            const latency: number = loggedOn - timestamp;
+
+            // passes into PostQuery class all query data along with recorded timestamp, loggedOn, and latency
+            const postQuery = new PostQuery(gateURI, projectID, {
```
Comment (shalarewicz) [inline src/index.ts]: What is the benefit of create a new `PostQuery` object? Are we able to just call a single `post` function as opposed to instantiated a PostQuery class and then calling post()?

File: src/middleware/PostQuery.ts [hunk 3]
```diff
@@ -21,57 +22,69 @@ export default class PostQuery {
     }
 
     // takes data from res.locals.graphqlGate and posts to webapp backend
-    public async post() {
-        const { complexity, timestamp, tokens, blocked } = this.queryData;
+    public async post(): Promise<void | Error> {
+        // needs depth property once added into limiter functionality
+        const { complexity, tokens, success, timestamp, loggedOn, latency } = this.queryData;
 
+        // most likely redundant, merely in place to fight possible edge cases
```
Comment (shalarewicz) [inline src/middleware/PostQuery.ts]: Agreed, it's redundant but can't hurt to check.

File: test/middleware/APIAuth.test.ts [hunk 4]
```diff
@@ -12,7 +12,7 @@ const MOCK_PROJECT_ID = '62997af7a5aab6a6df935797';
 const MOCK_API_KEY = 'Eo0sVUWQKM';
 const MOCK_URI = 'http://localhost:3000';
 
-describe('Test API key header verification', () => {
+xdescribe('Test API key header verification', () => {
```
Comment (shalarewicz) [inline test/middleware/APIAuth.test.ts]: Why is this test skipped?

File: test/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+                .expect('done')
+                .end((err, res) => {
+                    if (err) {
+                        return err;
+                    }
+                    return done();
+                });
+        });
+    });
+});
```
Comment (shalarewicz) [inline test/index.test.ts:87]: I think it's worthwhile to add tests for when bad data is passed in or a connection fails to make sure errors don't disrupt the middleware chain.

File: test/middleware/PostQuery.test.ts [hunk 6]
```diff
@@ -1,7 +1,8 @@
 import 'jest';
 import PostQuery from '../../src/middleware/PostQuery';
+import { QueryData } from '../../@types/log';
 
-describe('Test API key header verification', () => {
+xdescribe('Test API key header verification', () => {
```
Comment (shalarewicz) [inline test/middleware/PostQuery.test.ts]: Why is this test skipped?

File: test/server.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    };
+
+    res.locals.graphqlGate = MOCK_QUERY_DATA;
+
+    return next();
+});
+
+/**
+ * in theory, this res.send() should be calling not only res.end, but
+ * also the functionality added to res.end by the logger middleware
```
Comment (shalarewicz) [inline test/server.ts]: Update this comment to reflect the change to using an event listener?

**Review by shalarewicz — COMMENTED** [review]
Good work! Some minor comments but nothing major

File: src/index.ts (hunk 2 above)
Comment (jondeweydev) [reply src/index.ts]: changed to an exported function instead and tested, everything is working

File: src/index.ts (hunk 1 above)
Comment (jondeweydev) [reply src/index.ts]: changed it so that both of these functions either return true or an error, and if they do not return true, we return an error instead of returning the middleware callback

File: test/middleware/APIAuth.test.ts (hunk 4 above)
Comment (jondeweydev) [reply test/middleware/APIAuth.test.ts]: During testing of the index file, just so that testing would go by faster. Will change back for the final review

File: test/middleware/PostQuery.test.ts (hunk 6 above)
Comment (jondeweydev) [reply test/middleware/PostQuery.test.ts]: undone

File: test/server.ts (hunk 7 above)
Comment (jondeweydev) [reply test/server.ts]: done

File: test/index.test.ts (hunk 5 above)
Comment (jondeweydev) [reply test/index.test.ts:87]: will take care of this

**Review by feiw101 — APPROVED** [review]

File: src/index.ts [hunk 8]
```diff
@@ -19,42 +19,67 @@ const gateURI = 'http://localhost:3000';
...
-            // postQuery.post(req, res, next);
-            return res;
-        };
+    const newAuth = new AuthVerification(gateURI, projectID, apiKey);
+
+    const validate = newAuth.validation;
+    const verify = newAuth.verification;
+
+    // run the API Key verification process when gatelog is instantiated
+    validate()
```
Comment (evanmcneely) [inline src/index.ts]: This could be more DRY. Can you combine `validate` and `verify` into one function? Or use `Promise.all()` to run both functions and throw the error if one of them is `false`?

File: src/middleware/APIAuth.ts [hunk 9]
```diff
@@ -39,35 +39,41 @@ export default class AuthVerification {
...
     // Verifies key provided in header matches key in associated project's entry in DB
-    public async verification() {
+    // only returns true if api key is verified, otherwise returns error
+    public async verification(): Promise<boolean | Error> {
         let dbKey: string | Error = '';
 
         // this endpoint returns the associated project's API key
-        await axios(`${this.gateURI}/auth/${this.projectID}`)
+        await axios(`${this.gateURI}/key/${this.projectID}`)
             // .then((data: any) => data?.json())
```
Comment (evanmcneely) [inline src/middleware/APIAuth.ts]: You can remove this commented out `.then`

File: test/index.test.ts (hunk 5 above)
Comment (jondeweydev) [reply test/index.test.ts:87]: added

File: src/index.ts (hunk 8 above)
Comment (jondeweydev) [reply src/index.ts]: just saw this and took care of it

File: test/middleware/PostQuery.test.ts [hunk 10]
```diff
@@ -28,25 +30,24 @@ describe('Test API key header verification', () => {
...
+            loggedOn: 0,
         };
-        postQuery = new PostQuery(mockURI, mockProjectID, mockQueryData);
-        const newPost = () => postQuery.post();
+        const newPost = () => postQuery(mockURI, mockProjectID, mockQueryData);
 
         await expect(newPost).rejects.toThrow(`[gatelog] Query data cannot be negative.`);
     });
 
     test('Error thrown when resulting projectID of query does not match the one entered', async () => {
```
Comment (evanmcneely) [inline test/middleware/PostQuery.test.ts]: This test title reads like you are expecting an error to be thrown but then you're expecting an an error not to be thrown on line 44? What is the real intention?

File: test/middleware/PostQuery.test.ts [hunk 11]
```diff
@@ -28,25 +30,24 @@ describe('Test API key header verification', () => {
...
             timestamp: 100,
             tokens: -1,
+            success: true,
+            loggedOn: 0,
         };
-        postQuery = new PostQuery(mockURI, mockProjectID, mockQueryData);
-        const newPost = () => postQuery.post();
+        const newPost = () => postQuery(mockURI, mockProjectID, mockQueryData);
 
         await expect(newPost).rejects.toThrow(`[gatelog] Query data cannot be negative.`);
```
Comment (evanmcneely) [inline test/middleware/PostQuery.test.ts]: On my machine this test is failing because for the following reason
```
Received promise resolved instead of rejected
    Resolved to value: [SyntaxError: [gatelog] Query data cannot be negative]
```
It's resolving to the right error but not rejecting.

File: test/middleware/PostQuery.test.ts [hunk 12]
```diff
@@ -28,25 +30,24 @@ describe('Test API key header verification', () => {
...
-        postQuery = new PostQuery(mockURI, mockProjectID, mockQueryData);
-        const newPost = () => postQuery.post();
+        const newPost = () => postQuery(mockURI, mockProjectID, mockQueryData);
 
         await expect(newPost).not.toThrow(
             `[gatelog] GraphQL error, resulting query's projectID does not match the one entered`
         );
     });
 
     test('No error thrown when query data is correct syntax', async () => {
```
Comment (evanmcneely) [inline test/middleware/PostQuery.test.ts]: This is the same test as above but your but your not expecting any errors rather than a specific one. This test covers the above one, you don't need both.

File: test/middleware/PostQuery.test.ts [hunk 13]
```diff
@@ -1,11 +1,11 @@
...
+import { LogQueryData } from '../../@types/log';
 
 describe('Test API key header verification', () => {
     let mockURI: string;
     let mockProjectID: string;
-    let mockQueryData: QueryData;
-    let postQuery: PostQuery;
+    let mockQueryData: LogQueryData;
 
     beforeEach(() => {
```
Comment (evanmcneely) [inline test/middleware/PostQuery.test.ts]: Maybe this is why one of the tests below is failing for me. The tests should be agnostic to my local setup or move these to the `.env` file so I can set up the test environment on my machine without overriding these values for everyone.

Do I need to be running the web app on on Port 3000 to run the tests?

File: test/middleware/PostQuery.test.ts (hunk 10 above)
Comment (jondeweydev) [reply test/middleware/PostQuery.test.ts]: deleted this test

File: test/middleware/PostQuery.test.ts (hunk 11 above)
Comment (jondeweydev) [reply test/middleware/PostQuery.test.ts]: fixed

File: test/middleware/PostQuery.test.ts (hunk 12 above)
Comment (jondeweydev) [reply test/middleware/PostQuery.test.ts]: fixed

File: test/middleware/PostQuery.test.ts (hunk 13 above)
Comment (jondeweydev) [reply test/middleware/PostQuery.test.ts]: yes, these tests depend on running the webapp

**Comment by jondeweydev** [comment]
Fixed testing suite, all are passing and removed unnecessary tests. Feel free to test these out yourselves just keep in mind the web application's backend needs to be running for a majority of these to pass

**Comment by evanmcneely** [comment]
It looks like this PR may close issue #23 #17 #7 and #6 as well?

**Review by evanmcneely — APPROVED** [review]
Review issues associated with this PR. You may be able to link more issues and close them than is currently set up. 

Note: you need to put "closes" before each issue number for them to close automatically.

**Comment by jondeweydev** [comment]
closes #23 closes #16 closes #21 closes #17 closes #7 closes #6
